File: Programmers/1/92334. 신고 결과 받기/신고 결과 받기.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <unordered_set>

using namespace std;

vector<int> solution(vector<string> id_list, vector<string> report, int k) {
    vector<int> answer(id_list.size(), 0);
    
    unordered_set<string> unique(report.begin(), report.end());

    unordered_map<string, int> idIndex;
    
    for (int i = 0; i < (int)id_list.size(); ++i)
        idIndex[id_list[i]] = i;

    unordered_map<string, unordered_set<string>> reportedMap;

    for (const string& r : unique)
    {
        auto pos = r.find(' ');
        
        string reporter = r.substr(0, pos);
        string target = r.substr(pos + 1);
        
        reportedMap[target].insert(reporter);
    }
    
    for (auto& [target, reporters] : reportedMap)
    {
        if (reporters.size() >= k)
        {
            for (auto& reporter : reporters)
            {
                answer[idIndex[reporter]]++;
            }
        }
    }
    
    return answer;
}

File: Programmers/1/92334. 신고 결과 받기/신고 결과 받기.java
import java.util.*;

class Solution {
    public int[] solution(String[] id_list, String[] report, int k) {
        int[] answer = new int[id_list.length];
        
        Set<String> reportSet = new HashSet<>(Arrays.asList(report));
        Map<String, Set<String>> reportedMap = new HashMap<>();
        
        for (String r : reportSet) {
            String[] split = r.split(" ");
            String reporter = split[0];
            String target = split[1];
            
            reportedMap.putIfAbsent(target, new HashSet<>());
            reportedMap.get(target).add(reporter);
        }

        Set<String> suspended = new HashSet<>();
        
        for (Map.Entry<String, Set<String>> entry : reportedMap.entrySet()) {
            if (entry.getValue().size() >= k) {
                suspended.add(entry.getKey());
            }
        }

        Map<String, Integer> mailCount = new HashMap<>();
        
        for (String id : id_list) mailCount.put(id, 0);

        for (String target : suspended) {
            for (String reporter : reportedMap.get(target)) {
                mailCount.put(reporter, mailCount.get(reporter) + 1);
            }
        }

        for (int i = 0; i < id_list.length; i++) {
            answer[i] = mailCount.get(id_list[i]);
        }
        
        return answer;
    }
}